File: app/src/main/java/eu/darken/bluemusic/main/ui/managed/ManagedDevicesFragment.java
package eu.darken.bluemusic.main.ui.managed;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.core.view.ViewCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import eu.darken.bluemusic.R;
import eu.darken.bluemusic.bluetooth.ui.BluetoothActivity;
import eu.darken.bluemusic.main.core.audio.AudioStream;
import eu.darken.bluemusic.main.core.database.ManagedDevice;
import eu.darken.bluemusic.main.ui.MainActivity;
import eu.darken.bluemusic.main.ui.config.ConfigFragment;
import eu.darken.bluemusic.settings.ui.SettingsActivity;
import eu.darken.bluemusic.util.ActivityUtil;
import eu.darken.bluemusic.util.Check;
import eu.darken.mvpbakery.base.MVPBakery;
import eu.darken.mvpbakery.base.StateForwarder;
import eu.darken.mvpbakery.base.ViewModelRetainer;
import eu.darken.mvpbakery.injection.InjectedPresenter;
import eu.darken.mvpbakery.injection.PresenterInjectionCallback;


public class ManagedDevicesFragment extends Fragment implements ManagedDevicesPresenter.View, ManagedDevicesAdapter.Callback {

    @BindView(R.id.recyclerview) RecyclerView recyclerView;
    @BindView(R.id.empty_info) View emptyInfo;
    @BindView(R.id.bluetooth_disabled_container) View bluetoothDisabledContainer;
    @BindView(R.id.bluetooth_enabled_container) View bluetoothEnabledContainer;
    @BindView(R.id.fab) FloatingActionButton fab;

    @BindView(R.id.battery_saving_hint_container) View batterySavingHint;
    @BindView(R.id.battery_saving_hint_dismiss_action) Button batterySavingDismiss;
    @BindView(R.id.battery_saving_hint_show_action) Button batterySavingShow;

    @BindView(R.id.android10_applaunch_hint_container) View appLaunchHint;
    @BindView(R.id.android10_applaunch_hint_dismiss_action) Button appLaunchHintDismiss;
    @BindView(R.id.android10_applaunch_hint_show_action) Button appLaunchHintShow;
    @Inject ManagedDevicesPresenter presenter;

    private Unbinder unbinder;
    private ManagedDevicesAdapter adapter;
    private boolean isProVersion = false;
    private boolean bluetoothEnabled = false;
    private final StateForwarder stateForwarder = new StateForwarder();

    public static Fragment newInstance() {
        return new ManagedDevicesFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        stateForwarder.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View layout = inflater.inflate(R.layout.fragment_layout_managed_devices, container, false);
        unbinder = ButterKnife.bind(this, layout);
        return layout;
    }

    @SuppressLint("InlinedApi")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        adapter = new ManagedDevicesAdapter(this);
        recyclerView.setAdapter(adapter);
        ViewCompat.setNestedScrollingEnabled(recyclerView, false);

        bluetoothDisabledContainer.setOnClickListener(v -> presenter.showBluetoothSettingsScreen());

        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        stateForwarder.onSaveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onDestroyView() {
        if (unbinder != null) unbinder.unbind();
        super.onDestroyView();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        MVPBakery.<ManagedDevicesPresenter.View, ManagedDevicesPresenter>builder()
                .presenterFactory(new InjectedPresenter<>(this))
                .presenterRetainer(new ViewModelRetainer<>(this))
                .stateForwarder(stateForwarder)
                .addPresenterCallback(new PresenterInjectionCallback<>(this))
                .attach(this);

        super.onActivityCreated(savedInstanceState);
        Check.notNull(getActivity());
        final ActionBar actionBar = ((MainActivity) getActivity()).getSupportActionBar();
        Check.notNull(actionBar);
        actionBar.setDisplayHomeAsUpEnabled(false);
        actionBar.setTitle(R.string.app_name);
        actionBar.setSubtitle(null);
    }

    @Override
    public void onCreateOptionsMenu(@NotNull Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_main, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        menu.findItem(R.id.upgrade).setVisible(!isProVersion);

        Drawable btIcon = DrawableCompat.wrap(menu.findItem(R.id.bluetooth_settings).getIcon());
        int btStateColor = bluetoothEnabled ? android.R.color.white : R.color.state_m3;
        DrawableCompat.setTint(btIcon, ContextCompat.getColor(requireContext(), btStateColor));
        menu.findItem(R.id.bluetooth_settings).setIcon(btIcon);

        super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.upgrade:
                //noinspection ConstantConditions
                new AlertDialog.Builder(getContext())
                        .setTitle(R.string.label_premium_version)
                        .setMessage(R.string.description_premium_upgrade_explanation)
                        .setIcon(R.drawable.ic_stars_white_24dp)
                        .setPositiveButton(R.string.action_upgrade, (dialogInterface, i) -> presenter.onUpgradeClicked(getActivity()))
                        .setNegativeButton(R.string.action_cancel, (dialogInterface, i) -> {
                        })
                        .show();
                return true;
            case R.id.bluetooth_settings:
                presenter.showBluetoothSettingsScreen();
                return true;
            case R.id.settings:
                ActivityUtil.tryStartActivity(requireActivity(), new Intent(getContext(), SettingsActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @OnClick(R.id.fab)
    void onFabClicked() {
        ActivityUtil.tryStartActivity(this, new Intent(getContext(), BluetoothActivity.class));
    }

    @SuppressLint("RestrictedApi")
    @Override
    public void displayBluetoothState(boolean enabled) {
        bluetoothEnabled = enabled;
        bluetoothDisabledContainer.setVisibility(enabled ? View.GONE : View.VISIBLE);
        bluetoothEnabledContainer.setVisibility(enabled ? View.VISIBLE : View.GONE);
        fab.setVisibility(enabled ? View.VISIBLE : View.GONE);
        requireActivity().invalidateOptionsMenu();
    }

    @Override
    public void updateUpgradeState(boolean isProVersion) {
        this.isProVersion = isProVersion;
        //noinspection ConstantConditions
        getActivity().invalidateOptionsMenu();
    }

    @Override
    public void displayDevices(List<ManagedDevice> managedDevices) {
        adapter.setData(managedDevices);
        adapter.notifyDataSetChanged();

        recyclerView.setVisibility(managedDevices.isEmpty() ? View.GONE : View.VISIBLE);
        emptyInfo.setVisibility(managedDevices.isEmpty() ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onStreamAdjusted(ManagedDevice device, AudioStream.Type type, float percentage) {
        switch (type) {
            case MUSIC:
                presenter.onUpdateMusicVolume(device, percentage);
                break;
            case CALL:
                presenter.onUpdateCallVolume(device, percentage);
                break;
            case RINGTONE:
                presenter.onUpdateRingVolume(device, percentage);
                break;
            case NOTIFICATION:
                presenter.onUpdateNotificationVolume(device, percentage);
                break;
        }
    }

    @Override
    public void onShowConfigScreen(ManagedDevice device) {
        //noinspection ConstantConditions
        getActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.frame_content, ConfigFragment.instantiate(device))
                .addToBackStack(null)
                .commit();
    }

    @Override
    public void displayBatteryOptimizationHint(boolean display, Intent intent) {
        batterySavingDismiss.setOnClickListener(v -> presenter.onBatterySavingDismissed());
        batterySavingShow.setOnClickListener(v -> ActivityUtil.tryStartActivity(this, intent));
        batterySavingHint.setVisibility(display ? View.VISIBLE : View.GONE);
    }

    @Override
    public void displayAndroid10AppLaunchHint(boolean display, Intent intent) {
        appLaunchHintDismiss.setOnClickListener(v -> presenter.onAppLaunchHintDismissed());
        appLaunchHintShow.setOnClickListener(v -> ActivityUtil.tryStartActivity(this, intent));
        appLaunchHint.setVisibility(display ? View.VISIBLE : View.GONE);
    }
}

File: app/src/main/java/eu/darken/bluemusic/main/ui/managed/ManagedDevicesFragment.kt
package eu.darken.bluemusic.main.ui.managed

import android.annotation.SuppressLint
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.*
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.core.view.ViewCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import eu.darken.bluemusic.R
import eu.darken.bluemusic.bluetooth.ui.BluetoothActivity
import eu.darken.bluemusic.databinding.FragmentLayoutManagedDevicesBinding
import eu.darken.bluemusic.main.core.audio.AudioStream
import eu.darken.bluemusic.main.core.database.ManagedDevice
import eu.darken.bluemusic.main.ui.MainActivity
import eu.darken.bluemusic.main.ui.config.ConfigFragment.Companion.instantiate
import eu.darken.bluemusic.settings.ui.SettingsActivity
import eu.darken.bluemusic.util.ActivityUtil
import eu.darken.bluemusic.util.Check
import eu.darken.bluemusic.util.viewBinding
import eu.darken.mvpbakery.base.MVPBakery.Companion.builder
import eu.darken.mvpbakery.base.StateForwarder
import eu.darken.mvpbakery.base.ViewModelRetainer
import eu.darken.mvpbakery.injection.InjectedPresenter
import eu.darken.mvpbakery.injection.PresenterInjectionCallback
import javax.inject.Inject

class ManagedDevicesFragment : Fragment(), ManagedDevicesPresenter.View, ManagedDevicesAdapter.Callback {
    private val ui: FragmentLayoutManagedDevicesBinding by viewBinding()

    @Inject lateinit var presenter: ManagedDevicesPresenter
    private lateinit var adapter: ManagedDevicesAdapter
    private var isProVersion = false
    private var bluetoothEnabled = false
    private val stateForwarder = StateForwarder()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        stateForwarder.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_layout_managed_devices, container, false)
    }

    @SuppressLint("InlinedApi")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        ui.recyclerview.layoutManager = LinearLayoutManager(activity)
        adapter = ManagedDevicesAdapter(this)
        ui.recyclerview.adapter = adapter
        ViewCompat.setNestedScrollingEnabled(ui.recyclerview, false)
        ui.bluetoothDisabledContainer.setOnClickListener { v: View? -> presenter.showBluetoothSettingsScreen() }

        ui.fab.setOnClickListener {
            ActivityUtil.tryStartActivity(this, Intent(context, BluetoothActivity::class.java))
        }

        super.onViewCreated(view, savedInstanceState)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        stateForwarder.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        builder<ManagedDevicesPresenter.View, ManagedDevicesPresenter>()
                .presenterFactory(InjectedPresenter(this))
                .presenterRetainer(ViewModelRetainer(this))
                .stateForwarder(stateForwarder)
                .addPresenterCallback(PresenterInjectionCallback(this))
                .attach(this)
        super.onActivityCreated(savedInstanceState)
        Check.notNull(activity)
        val actionBar = (requireActivity() as MainActivity).supportActionBar!!
        actionBar.setDisplayHomeAsUpEnabled(false)
        actionBar.setTitle(R.string.app_name)
        actionBar.subtitle = null
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_main, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        menu.findItem(R.id.upgrade).isVisible = !isProVersion
        val btIcon = DrawableCompat.wrap(menu.findItem(R.id.bluetooth_settings).icon!!)
        val btStateColor = if (bluetoothEnabled) android.R.color.white else R.color.state_m3
        DrawableCompat.setTint(btIcon, ContextCompat.getColor(requireContext(), btStateColor))
        menu.findItem(R.id.bluetooth_settings).icon = btIcon
        super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.upgrade -> {
                AlertDialog.Builder(requireContext())
                        .setTitle(R.string.label_premium_version)
                        .setMessage(R.string.description_premium_upgrade_explanation)
                        .setIcon(R.drawable.ic_stars_white_24dp)
                        .setPositiveButton(R.string.action_upgrade) { dialogInterface: DialogInterface?, i: Int -> presenter.onUpgradeClicked(activity) }
                        .setNegativeButton(R.string.action_cancel) { dialogInterface: DialogInterface?, i: Int -> }
                        .show()
                true
            }
            R.id.bluetooth_settings -> {
                presenter.showBluetoothSettingsScreen()
                true
            }
            R.id.settings -> {
                ActivityUtil.tryStartActivity(requireActivity(), Intent(context, SettingsActivity::class.java))
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    @SuppressLint("RestrictedApi") override fun displayBluetoothState(enabled: Boolean) {
        bluetoothEnabled = enabled
        ui.bluetoothDisabledContainer.visibility = if (enabled) View.GONE else View.VISIBLE
        ui.bluetoothEnabledContainer.visibility = if (enabled) View.VISIBLE else View.GONE
        ui.fab.visibility = if (enabled) View.VISIBLE else View.GONE
        requireActivity().invalidateOptionsMenu()
    }

    override fun updateUpgradeState(isProVersion: Boolean) {
        this.isProVersion = isProVersion
        requireActivity().invalidateOptionsMenu()
    }

    override fun displayDevices(managedDevices: List<ManagedDevice>) {
        adapter.data = managedDevices
        adapter.notifyDataSetChanged()
        ui.recyclerview.visibility = if (managedDevices.isEmpty()) View.GONE else View.VISIBLE
        ui.emptyInfo.visibility = if (managedDevices.isEmpty()) View.VISIBLE else View.GONE
    }

    override fun onStreamAdjusted(device: ManagedDevice, type: AudioStream.Type, percentage: Float) {
        when (type) {
            AudioStream.Type.MUSIC -> presenter.onUpdateMusicVolume(device, percentage)
            AudioStream.Type.CALL -> presenter.onUpdateCallVolume(device, percentage)
            AudioStream.Type.RINGTONE -> presenter.onUpdateRingVolume(device, percentage)
            AudioStream.Type.NOTIFICATION -> presenter.onUpdateNotificationVolume(device, percentage)
            AudioStream.Type.ALARM -> presenter.onUpdateAlarmVolume(device, percentage)
        }
    }

    override fun onShowConfigScreen(device: ManagedDevice) {
        requireActivity().supportFragmentManager.beginTransaction()
                .replace(R.id.frame_content, instantiate(device))
                .addToBackStack(null)
                .commit()
    }

    override fun displayBatteryOptimizationHint(display: Boolean, intent: Intent) {
        ui.batterySavingHintDismissAction.setOnClickListener { v: View? -> presenter.onBatterySavingDismissed() }
        ui.batterySavingHintShowAction.setOnClickListener { v: View? -> ActivityUtil.tryStartActivity(this, intent) }
        ui.batterySavingHintContainer.visibility = if (display) View.VISIBLE else View.GONE
    }

    override fun displayAndroid10AppLaunchHint(display: Boolean, intent: Intent) {
        ui.android10ApplaunchHintDismissAction.setOnClickListener { v: View? -> presenter.onAppLaunchHintDismissed() }
        ui.android10ApplaunchHintShowAction.setOnClickListener { v: View? -> ActivityUtil.tryStartActivity(this, intent) }
        ui.android10ApplaunchHintContainer.visibility = if (display) View.VISIBLE else View.GONE
    }

    companion object {
        @JvmStatic fun newInstance(): Fragment {
            return ManagedDevicesFragment()
        }
    }
}